### PR TITLE
APIRule hosts should be FQDN

### PR DESCRIPTION
**Pull request by mluk-sap**
**Description**

Changes proposed in this pull request:
- in APIRule hosts are expected to be in FQDN form, added proper validation and tests

**Pre-Merge Checklist**
- [x] As a PR reviewer, verify code coverage and evaluate if it is acceptable.

**Related issues**

### CODE REVIEW

File: internal/validation/v2alpha1/hosts_test.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+		//when
+		problems := validateHosts(".spec", networkingv1beta1.VirtualServiceList{}, apiRule)
+
+		//then
+		Expect(problems).To(HaveLen(1))
+		Expect(problems[0].AttributePath).To(Equal(".spec.hosts[1]"))
+		Expect(problems[0].Message).To(Equal("Host is not fully qualified domain name"))
+	})
+
```
Comment (videlov) [inline internal/validation/v2alpha1/hosts_test.go:165]: please add/cover all tests for the length limitations on FQDN validation:
```

const (
	fqdnMaxLength        = 253
	fqdnMinSegments      = 2
	domainLabelMaxLength = 63
)
```

File: internal/validation/v2alpha1/hosts_test.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+		//when
+		problems := validateHosts(".spec", virtualServiceList, apiRule)
+
+		//then
+		Expect(problems).To(HaveLen(1))
+		Expect(problems[0].AttributePath).To(Equal(".spec.hosts[1]"))
+		Expect(problems[0].Message).To(Equal("Host is occupied by another Virtual Service"))
+	})
+
+	It("Should not fail if host is occupied by the Virtual Service exposed by another resource", func() {
```
Comment (videlov) [inline internal/validation/v2alpha1/hosts_test.go:234]: this is same test as the one above?

**Review by videlov — CHANGES_REQUESTED** [review]

File: internal/validation/v2alpha1/hosts_test.go (hunk 2 above)
Comment (mluk-sap) [reply internal/validation/v2alpha1/hosts_test.go:234]: Ah, right, copy/paste issue...

File: internal/validation/v2alpha1/hosts_test.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+		//when
+		problems := validateHosts(".spec", virtualServiceList, apiRule)
+
+		//then
+		Expect(problems).To(HaveLen(1))
+		Expect(problems[0].AttributePath).To(Equal(".spec.hosts[1]"))
+		Expect(problems[0].Message).To(Equal("Host is occupied by another Virtual Service"))
+	})
+
+	It("Should not fail if a host is occupied by the Virtual Service related to the same API Rule", func() {
```
[the remark this answers is not in the document]
Comment (mluk-sap) [reply internal/validation/v2alpha1/hosts_test.go:275]: It is already there under "Should fail for any host that is occupied by any Virtual Service exposed by another resource".
But strictly speaking - it just checks for a Virtual Service with the same host, which may be even not linked to any APIRule (like the one created manually by the end-user)

File: internal/validation/v2alpha1/hosts.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			hostAttributePath := fmt.Sprintf("%s[%d]", hostsAttributePath, hostIndex)
+			failures = append(failures, validation.Failure{
+				AttributePath: hostAttributePath,
+				Message:       "Host is not fully qualified domain name",
+			})
+		}
+	}
+
+	for _, vs := range vsList.Items {
+		for hostIndex, host := range hosts {
```
[the remark this answers is not in the document]
Comment (mluk-sap) [reply internal/validation/v2alpha1/hosts.go]: done

File: internal/validation/v2alpha1/hosts.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		return false
+	}
+
+	labels := strings.Split(host, ".")
+	if len(labels) < fqdnMinSegments {
+		return false
+	}
+
+	for _, domainLabel := range labels {
+		if len(domainLabel) == 0 || len(domainLabel) > domainLabelMaxLength {
```
[the remark this answers is not in the document]
Comment (mluk-sap) [reply internal/validation/v2alpha1/hosts.go]: done

File: internal/validation/v2alpha1/hosts_test.go (hunk 1 above)
Comment (mluk-sap) [reply internal/validation/v2alpha1/hosts_test.go:167]: I've just added much more tests covering all these cases

File: internal/validation/v2alpha1/hosts_test.go [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+		//when
+		problems := validateHosts(".spec", virtualServiceList, apiRule)
+
+		//then
+		Expect(problems).To(HaveLen(1))
+		Expect(problems[0].AttributePath).To(Equal(".spec.hosts[0]"))
+		Expect(problems[0].Message).To(Equal("Host is occupied by another Virtual Service"))
+	})
+
+	It("Should fail for any host that is occupied by any Virtual Service exposed by another resource", func() {
```
[the remark this answers is not in the document]
Comment (mluk-sap) [reply internal/validation/v2alpha1/hosts_test.go]: yes, I've just removed it

File: internal/validation/v2alpha1/hosts_test.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+		//then
+		Expect(problems).To(HaveLen(1))
+		Expect(problems[0].AttributePath).To(Equal(".spec.hosts[1]"))
+		Expect(problems[0].Message).To(Equal("Host is occupied by another Virtual Service"))
+	})
+
+	It("Should not fail if a host is occupied by the Virtual Service related to the same API Rule", func() {
+		//given
+		apiRule := &v2alpha1.APIRule{
```
Comment (videlov) [inline internal/validation/v2alpha1/hosts_test.go:282]:
```suggestion
		apiRule := &v2alpha1.APIRule{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "some-name",
				Namespace: "some-ns",
			},
```

Comment (videlov) [reply internal/validation/v2alpha1/hosts_test.go:282]: just for the same of `getExpectedOwnerLabel` thats building a string out of the metadata

File: internal/validation/v2alpha1/hosts_test.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+		Expect(problems).To(HaveLen(4))
+		for i := 0; i < 4; i++ {
+			Expect(problems[i].AttributePath).To(Equal(fmt.Sprintf(".spec.hosts[%d]", i)))
+			Expect(problems[i].Message).To(Equal("Host is not fully qualified domain name"))
+		}
+	})
+
+	It("Should fail if any host that is occupied by any Virtual Service exposed by another resource", func() {
+		//given
+		apiRule := &v2alpha1.APIRule{
```
Comment (videlov) [inline internal/validation/v2alpha1/hosts_test.go:241]:
```suggestion
		apiRule := &v2alpha1.APIRule{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "some-name",
				Namespace: "some-ns",
			},
```

Comment (videlov) [reply internal/validation/v2alpha1/hosts_test.go:241]: just for the same of `getExpectedOwnerLabel` thats building a string out of the metadata

**Review by videlov — CHANGES_REQUESTED** [review]

**Review by videlov — APPROVED** [review]

File: docs/user/custom-resources/apirule/v2alpha1/04-10-apirule-custom-resource.md [hunk 9]
```diff
@@ -23,7 +23,7 @@ This table lists all parameters of APIRule `v2alpha1` CRD together with their de
...
-| **hosts**                                        |  **YES**  | Specifies the Service's communication address for inbound external traffic. If only the leftmost label is provided, the default domain name is used.                                                                                                                                  | The full domain name or the leftmost label cannot contain the wildcard character `*`.                 |
+| **hosts**                                        |  **YES**  | Specifies the Service's communication address for inbound external traffic. It is expected to be a fully qualified domain name.                                                                                                                                                       | It should be proper FQDN format: at least 2 domain labels with characters, numbers or dashes.         |
```
Comment (nataliasitko) [inline docs/user/custom-resources/apirule/v2alpha1/04-10-apirule-custom-resource.md:26]:
```suggestion
| **hosts**                                        |  **YES**  | Specifies the Service's communication address for inbound external traffic. It must be a fully qualified domain name in proper FQDN format: at least two domain labels with characters, numbers, or dashes.         |
```

**Review by nataliasitko — APPROVED** [review]

**Review by videlov — APPROVED** [review]

**Pull request merged**